<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/15-puti-cerez-tablicu' style="text-decoration: none; color: #1a73e8;">
            15-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 15

Путь начинается в левом верхнем углу сетки. На каждом шаге разрешено двигаться только:

- вправо;
- вниз.

Для сетки размером `2 × 2` существует ровно `6` различных маршрутов до правого нижнего угла.

<img src="../Sources/Project_Euler_015.png" width="300">

Необходимо определить, сколько таких маршрутов существует в сетке размером `20 × 20`.

Ответ: 137846528820

## 1) Динамическое программирование

В данном решении используется метод динамического программирования.

Пусть `s[i][j]` — количество способов попасть в клетку `(i, j)`. Поскольку двигаться можно только вправо и вниз, в каждую клетку можно попасть либо сверху, либо слева.

Следовательно, количество путей вычисляется по формуле:

$$
s[i][j] = s[i - 1][j] + s[i][j - 1]
$$

Первая строка и первый столбец содержат только один возможный путь, поэтому они заранее заполняются единицами.

### Алгоритм

1. Создаётся таблица размером `(n + 1) × (n + 1)`.
2. Первая строка и первый столбец заполняются единицами.
3. Последовательно вычисляются значения остальных клеток.
4. Для каждой клетки суммируются значения сверху и слева.
5. После заполнения всей таблицы ответ находится в правом нижнем углу.

### Особенности

- Используется динамическое программирование.
- Каждая клетка вычисляется только один раз.
- Для хранения промежуточных результатов используется двумерная таблица.
- Временная сложность составляет `O(n²)`.
- Пространственная сложность составляет `O(n²)`.

### Вывод

Динамическое программирование позволяет эффективно вычислить количество путей, используя уже найденные значения соседних клеток. Благодаря этому каждая подзадача решается только один раз, что делает алгоритм значительно быстрее полного перебора всех маршрутов.

In [13]:
from time import perf_counter # измерение времени выполнения алгоритмов

In [14]:
def get_result_1(n):
    n += 1

    s = [[0] * n for i in range(n)]

    for i in range(n):
        s[0][i] = 1
        s[i][0] = 1


    for i in range(1, n):
        for j in range(1, n):
            s[i][j] = s[i - 1][j] + s[i][j - 1]
    
    return s[n - 1][n - 1]

        
start = perf_counter()
print(get_result_1(20))
end = perf_counter()

print(f'{end - start:.5f}')

137846528820
0.00012


## 2) Динамическое программирование с одномерным массивом

Во втором решении используется та же идея динамического программирования, однако вместо двумерной таблицы применяется одномерный массив.

Элемент `s[i]` хранит количество путей до текущей клетки текущей строки. При переходе к следующей строке значения массива последовательно обновляются по формуле:

```python
s[i] += s[i - 1]
```

В результате значение сверху уже находится в `s[i]`, а значение слева — в `s[i - 1]`.

### Алгоритм

1. Создаётся одномерный массив, заполненный единицами.
2. Последовательно обрабатывается каждая строка таблицы.
3. Для каждого элемента, начиная со второго, выполняется обновление:
   ```python
   s[i] += s[i - 1]
   ```
4. После обработки всех строк последний элемент массива содержит искомое количество маршрутов.

### Отличия от предыдущего решения

- Вместо двумерной таблицы используется одномерный массив.
- Значения обновляются непосредственно в процессе обхода строк.
- Дополнительная память уменьшается с `O(n²)` до `O(n)`.
- Временная сложность остаётся `O(n²)`.

### Вывод

Использование одномерного массива позволяет существенно сократить потребление памяти без изменения идеи алгоритма. Благодаря обновлению значений "на месте" решение остаётся таким же быстрым, но становится более экономичным.

In [15]:
def get_result_2(n):
    n += 1
    s = [1] * n

    for _ in range(1, n):
        for i in range(1, n):
            s[i] += s[i - 1]

    return s[n - 1]

        
start = perf_counter()
print(get_result_2(20))
end = perf_counter()

print(f'{end - start:.5f}')

137846528820
0.00008


## 3) Использование биномиального коэффициента

В данном решении задача сводится к вычислению биномиального коэффициента.
Чтобы попасть из левого верхнего угла в правый нижний угол сетки `n × n`, необходимо сделать ровно `2n` шагов: `n` вправо и `n` вниз.
Таким образом, задача сводится к подсчёту количества способов выбрать позиции для одного из двух типов шагов.
Количество различных маршрутов равно биномиальному коэффициенту:

$$
C_{2n}^{n}
$$

Для вычисления используется встроенная функция `comb()` из модуля `math`, которая эффективно находит значение сочетания без вычисления факториалов целиком.

### Алгоритм

1. Вычисляется общее количество шагов `2n`.
2. Определяется число шагов одного направления (`n`).
3. Вычисляется биномиальный коэффициент с помощью функции `comb()`.
4. Возвращается полученный результат.

### Отличия от предыдущих решений

- Динамическое программирование не используется.
- Задача сводится к вычислению одного биномиального коэффициента.
- Для вычислений применяется функция `comb()`.
- Дополнительная память практически не используется.
- Временная сложность близка к `O(1)` для фиксированного значения `n`.

### Вывод

Использование биномиального коэффициента позволяет решить задачу наиболее элегантным способом. Вместо построения таблицы динамического программирования вычисляется готовая комбинаторная формула, благодаря чему решение получается одновременно самым коротким и самым быстрым.

In [16]:
from math import comb, ceil


def get_result_3(n):
    n *= 2
    k = ceil(n / 2)
    return comb(n, k)
    

start = perf_counter()
print(get_result_3(20))
end = perf_counter()

print(f'{end - start:.5f}')

137846528820
0.00005


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **1000 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [17]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(20), number=1000) / 1000
    print(f'Алгоритм {i}: {tm:.10f}')

Алгоритм 1: 0.0000330522
Алгоритм 2: 0.0000191619
Алгоритм 3: 0.0000001383


## Заключение

В ходе решения задачи были рассмотрены три различных подхода к подсчёту количества маршрутов в квадратной сетке.

Первый алгоритм использует классическое динамическое программирование с двумерной таблицей. Для каждой клетки вычисляется количество путей как сумма значений сверху и слева. Такой подход прост для понимания, однако требует хранения всей таблицы.

Во втором решении используется та же идея динамического программирования, но вместо двумерного массива применяется одномерный. Значения обновляются непосредственно в процессе обхода строк, благодаря чему потребление памяти уменьшается с `O(n²)` до `O(n)`, а время выполнения также немного сокращается.

Третий алгоритм основан на комбинаторике. Вместо последовательного вычисления количества путей используется тот факт, что любой маршрут состоит из `n` шагов вправо и `n` шагов вниз. Поэтому ответ равен биномиальному коэффициенту

$$
C_{2n}^{n},
$$

который вычисляется встроенной функцией `comb()`.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `0.0000330522` с
- **Алгоритм 2:** `0.0000191619` с
- **Алгоритм 3:** `0.0000001383` с

Полученные результаты показывают, что оптимизация структуры данных позволяет ускорить динамическое программирование примерно в **1.7 раза**. Однако наибольший выигрыш достигается при переходе к математической модели задачи. Использование биномиального коэффициента позволило отказаться от построения таблицы и сделать решение более чем в **130 раз быстрее** первой реализации, одновременно получив самый компактный и элегантный код.